radicle-httpd/src/raw.rs

```diff
@@ -0,0 +1,54 @@
+use std::sync::Arc;
+
+use axum::extract::State;
+use axum::response::IntoResponse;
+use axum::routing::get;
+use axum::Router;
+
+use radicle::prelude::Id;
+use radicle::profile::Profile;
+use radicle::storage::git::paths;
+use radicle_surf::{Oid, Repository};
+
+use crate::api::axum_extra::Path;
+use crate::api::error::Error;
+
+pub fn router(profile: Arc<Profile>) -> Router {
+    Router::new()
+        .route("/:project/:sha/*path", get(file_handler))
+        .with_state(profile)
+}
+
+async fn file_handler(
+    Path((project, sha, path)): Path<(Id, Oid, String)>,
+    State(profile): State<Arc<Profile>>,
+) -> impl IntoResponse {
+    let storage = &profile.storage;
+    let repo = Repository::open(paths::repository(storage, &project))?;
+    let blob = repo.blob(sha, &path)?;
+
```

Comment:
`blob` is fully allocated, a more efficient way would be to implement a stream, any thoughts @keepsimple1 ?

Reply:
btw, this causes file to be downloaded, not shown in browser. good enough for me, lmk if you want it changed.

Reply:
I think if we could include `repo` inside `State(profile)`,  so that we don't have to (re)open the repo every time inside the handler, that might help. 

Since `blob<'a>` has its lifetime associated with `repo`, such change will allow us return a `Blob` without calling `to_owned()`, hence would be helpful for streaming.

I could be wrong but this link seems to show it might be possible to include `repo` in the profile? https://github.com/radicle-dev/heartwood/blob/2b5552546c6dc637aaa282a990083df8a4a0c0c0/radicle-httpd/src/api.rs#L49-L52

Reply:
> I think if we could include repo inside State(profile), so that we don't have to (re)open the repo every time inside the handler, that might help.

I'm not sure that's a good idea. You would be carrying around a handle to the `Repository` the whole time, which isn't always what you want.

Reply:
The file should be shown in the browser if it's non-binary, which requires the correct content-type to be set I think.

Reply:
@keepsimple1 not sure what you mean, but can't you just `Box` whatever you need?

Reply:
> The file should be shown in the browser if it's non-binary, which requires the correct content-type to be set I think.

if I set `header::CONTENT_TYPE, "text; charset=utf-8"` then:
- text files would be shown as text ✓
- images as images ✓
- binary files would start a download ✓

it shouldn't work, but browsers are doing the heavy lifting to detect the mime and works, otherwise if we want the correct mime we have to add another crate. lmk.

Reply:
> @keepsimple1 not sure what you mean, but can't you just `Box` whatever you need?

The purpose of my suggestion is to reduce memory usage of large blobs. 

I meant that if we could return `blob.content()` from this method, we will have a `&[u8]` that is provided by the underlying `git2::blob`. We will not use extra memory. 

For `&[u8]`,  we could stream from it if `heartwood` already have such facility. Or, add a method to return a chunk (slice) via an iterator. 

> can't you just `Box` whatever you need?

I don't think returning a `Box<&[u8]>` could not really help as the issue is the lifetime is associated with the underlying git2 repo. If we want to return `Box<Vec<u8>>`, then we will need to first make a copy to produce `Vec<u8>` anyway, which will consume extra memory.

Another approach is to write large blobs into a temp file, and then streaming from the temp file.  At the end, delete the temp file.